### PR TITLE
CFE-3048: Default permissions for created files

### PR DESCRIPTION
**Pull request by karlhto**
Also added default permissions for safe_open, with corresponding
constants in definitions.h

Going to continue by rectifying all uses of `safe_fopen`.

Ticket: CFE-3048
Changelog: None

### CODE REVIEW

**Comment by karlhto** [comment]
@vpodzime I am unsure if doing it this way is best practice. What do you think?

**Comment by karlhto** [comment]
@cf-bottom Jenkins please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2337/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2337/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2337/

File: libutils/file_lib.c [hunk 1]
```diff
@@ -835,7 +835,7 @@ FILE *safe_fopen(const char *path, const char *mode, ...)
     }
     else
     {
-        fd = safe_open(path, flags);
+        fd = safe_open(path, flags, 0);
```
[the remark this answers is not in the document]
Comment (karlhto) [reply libutils/file_lib.c]: LGTM thought it was good practice to be explicit about it, though I am still on the fence about whether we should use variadic functions at all.

[the remark this answers is not in the document]
Comment (vpodzime) [reply libutils/file_lib.c]: > I am still on the fence about whether we should use variadic functions at all.

I think it makes sense due to the standard `open()` doing the same. But let's see what @olehermanse thinks.

**Comment by karlhto** [comment]
After giving it some more thought, I think using the same flags as `open` is a better idea, since it is much less ambiguous. I also want to use `safe_fopen` with those arguments in the new functions in `libcfnet`.

File: libutils/file_lib.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (olehermanse) [reply libutils/file_lib.c]: @vpodzime @karlhto It makes sense to me that `safe_open()` accepts the same arguments as `open()`, and `safe_fopen()` matches `fopen()`. 

When it comes to the optional argument for `open()`, I really don't like it. I think I (slightly) prefer to have 2 functions, `safe_open()` taking 2 arguments only, and defaulting to restrictive `mode` argument on `O_CREAT`. `safe_open_mode()` (or similar name, `safe_open_perms()`) can have the required, third, `mode` parameter. The implementation of `safe_open()` would then call `safe_open_mode()` internally.

[the remark this answers is not in the document]
Comment (karlhto) [reply libutils/file_lib.c]: @olehermanse I agree, seems like that's the best way to do it in C -- especially because it might seem kind of arbitrary in which situations you are supposed to use the optional argument.

I think that `safe_open` (and then in effect `safe_fopen`) should default to mode `0600`, and then we should have additional functions (like `safe_open_mode` and `safe_fopen_mode`) if you explicitly want to define a different mode.

**Comment by karlhto** [comment]
@olehermanse @vpodzime So, do we agree that we should split it into two functions? And is it only `safe_fopen` that should be split into two, or `safe_open` as well? `safe_open` already had support for variable mode arg.

**Comment by olehermanse** [comment]
@karlhto I believe the [JIRA ticket description](https://tracker.mender.io/browse/CFE-3048) reflects our latest "consensus", in short:

* No valist
    * Separate `_perms` functions for specifying perms on create. `create_perms` parameter name for clarity
* Don't check umask - it will only restrict the default or explicit perms further
* Default perms 0600 for file creations in both functions

File: libutils/definitions.h [hunk 2]
```diff
@@ -44,4 +44,11 @@
 #define CF_BUFSIZE     4096
 #define CF_EXPANDSIZE (2 * CF_BUFSIZE)
 
+
+/*****************************************************************************
+ *                          File permissions                                 *
+ *****************************************************************************/
+#define CF_MODE_DEFAULT  S_IWUSR | S_IRUSR
+#define CF_MODE_SHARED   CF_MODE_DEFAULT | S_IRGRP | S_IROTH
```
Comment (olehermanse) [inline libutils/definitions.h]: Since mode means something different for `open()` and `fopen()` I would prefer something more clear/explicit:

```C
// 600 - Read/Write for owner:
#define CF_CREATE_PERMS_DEFAULT S_IWUSR | S_IRUSR
// 644 - World readable:
#define CF_CREATE_PERMS_WORLD_READABLE CF_CREATE_PERMS_DEFAULT | S_IRGRP | S_IROTH
```

Comment (karlhto) [reply libutils/definitions.h]: In the branch in my fork I changed them to `CF_PERMS_DEFAULT` and `CF_PERMS_SHARED`. I think it might be better to use `S_IWUSR` and similar instead, though.

**Comment by karlhto** [comment]
@olehermanse Sounds good. Could you throw a WIP on this?

**Comment by karlhto** [comment]
@cf-bottom Jenkins please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2411/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2411/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2411/

File: libpromises/bootstrap.c [hunk 3]
```diff
@@ -269,7 +269,7 @@ bool WriteBuiltinFailsafePolicyToPath(const char *filename)
 
     Log(LOG_LEVEL_INFO, "Writing built-in failsafe policy to '%s'", filename);
 
-    FILE *fout = fopen(filename, "w");
+    FILE *fout = safe_fopen_create_perms(filename, "w", CF_PERMS_DEFAULT);
```
Comment (vpodzime) [inline libpromises/bootstrap.c]: We can now drop the `chmod()` code below, right?

File: cf-agent/verify_files_hashes.c [hunk 4]
```diff
@@ -82,8 +82,8 @@ bool CompareBinaryFiles(const char *file1, const char *file2, const struct stat
 
     if (conn == NULL)
     {
-        fd1 = safe_open(file1, O_RDONLY | O_BINARY, 0400);
-        fd2 = safe_open(file2, O_RDONLY | O_BINARY, 0400);
+        fd1 = safe_open(file1, O_RDONLY | O_BINARY);
+        fd2 = safe_open(file2, O_RDONLY | O_BINARY);
```
Comment (olehermanse) [inline cf-agent/verify_files_hashes.c]: 0400 is not the default, right? So why did you remove it instead of changing to safe_open_create_perms?

File: cf-serverd/cf-serverd-functions.c [hunk 5]
```diff
@@ -121,7 +121,7 @@ static int GenerateAvahiConfig(const char *path)
 {
     FILE *fout;
     Writer *writer = NULL;
-    fout = safe_fopen(path, "w+");
+    fout = safe_fopen_create_perms(path, "w+", CF_PERMS_DEFAULT);
```
Comment (olehermanse) [inline cf-serverd/cf-serverd-functions.c]: Why not just use safe_fopen(), which would use the default perms?

File: cf-agent/verify_files_utils.c [hunk 6]
```diff
@@ -3914,7 +3914,8 @@ bool CfCreateFile(EvalContext *ctx, char *file, const Promise *pp, const Attribu
 
             MakeParentDirectory(file, attr->move_obstructions);
 
-            int fd = safe_open(file, O_WRONLY | O_CREAT | O_EXCL, filemode);
+            int fd = safe_open_create_perms(file, O_WRONLY | O_CREAT | O_EXCL,
+                                            filemode);
```
Comment (olehermanse) [inline cf-agent/verify_files_utils.c]: Could be on one line.

File: cf-key/cf-key-functions.c [hunk 7]
```diff
@@ -232,19 +232,13 @@ int RemoveKeys(const char *input, bool must_be_coherent)
...
-
-    if (fd < 0)
-    {
-        Log(LOG_LEVEL_ERR, "Couldn't open private key file '%s' (open: %s)", private_key_file, GetErrorStr());
-        return false;
-    }
-
-    FILE *fp = fdopen(fd, "w");
+    FILE *fp = safe_fopen_create_perms(private_key_file, "w",
+                                       CF_PERMS_DEFAULT);
```
Comment (olehermanse) [inline cf-key/cf-key-functions.c]: As @vpodzime pointed out, private keys should not use the default perms macro, even though they are currently corrrect. Either make a separate macro `CF_PERMS_PRIVKEY`, or keep the hard coded values.

**Comment by olehermanse** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2449/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2449/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2449/

File: cf-agent/verify_files_hashes.c (hunk 4 above)
Comment (olehermanse) [reply cf-agent/verify_files_hashes.c]: I guess because there is no `O_CREAT`.

**Comment by olehermanse** [comment]
@cf-bottom jenkins please

**Comment by olehermanse** [comment]
@vpodzime I tried to address all review comments here in one final commit. Please review the last commit so we can get this merged soon.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2504/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2504/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2504/

File: libutils/file_lib.c [hunk 8]
```diff
@@ -491,28 +491,27 @@ int safe_open(const char *pathname, int flags)
...
+ * safe_open and safe_fopen both default to secure (0600) file creation perms.
+ * The _create_perms variants allow you to explicitly set different permissions.
  *
- * @param pathname The path to open.
- * @param flags Same flags as for system open().
- * @param perms Permissions for file, only relevant on file creation.
- * @return Same errors as open().
+ * @param pathname The path to open
+ * @param flags Same flags as for system open()
+ * @param perms Permissions for file, only relevant on file creation
```
Comment (vpodzime) [inline libutils/file_lib.c]: Should be `@param create_perms` now.

File: libutils/file_lib.c [hunk 9]
```diff
@@ -792,13 +791,11 @@ FILE *safe_fopen(const char *const path, const char *const mode)
  *
  * @param pathname The path to open.
  * @param flags Same mode as for system fopen().
- * @param ... Optional permissions argument, same as in safe_open()
+ * @param perms Permissions for file, only relevant on file creation.
```
Comment (vpodzime) [inline libutils/file_lib.c]: Same here.

**Comment by olehermanse** [comment]
@cf-bottom jenkins, again!

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2507/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2507/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2507/
